File: content/17-math-and-geometry/large_factorial.js
/**
 * Problem: Factorials of Large Numbers
 * GeeksforGeeks: https://practice.geeksforgeeks.org/problems/factorials-of-large-numbers/0
 */

function multiply(x, res) {
    let carry = 0;
    for (let i = 0; i < res.length; i++) {
        const prod = res[i] * x + carry;
        res[i] = prod % 10;
        carry = Math.floor(prod / 10);
    }
    
    while (carry > 0) {
        res.push(carry % 10);
        carry = Math.floor(carry / 10);
    }
}

function factorial(n) {
    /**
     * Time Complexity: O(N^2)
     * Space Complexity: O(N)
     */
    const res = [1];
    for (let x = 2; x <= n; x++) {
        multiply(x, res);
    }
    return res.reverse();
}

if (require.main === module) {
    const assert = require('assert');
    assert.strictEqual(factorial(5).join(''), "120");
    assert.strictEqual(factorial(10).join(''), "3628800");
    console.log("JS: factorial passed tests!");
}

File: content/17-math-and-geometry/large_factorial.py
"""
Problem: Factorials of Large Numbers
GeeksforGeeks: https://practice.geeksforgeeks.org/problems/factorials-of-large-numbers/0
"""

def multiply(x: int, res: list[int]) -> int:
    """
    Multiplies x with the number represented by res array and updates res.
    Returns the new size of res.
    """
    carry = 0
    for i in range(len(res)):
        prod = res[i] * x + carry
        res[i] = prod % 10
        carry = prod // 10
        
    while carry:
        res.append(carry % 10)
        carry //= 10
        
    return len(res)

def factorial(n: int) -> list[int]:
    """
    Calculates factorial of a large number and returns it as a list of digits (reversed).
    Time Complexity: O(N^2)
    Space Complexity: O(N)
    """
    res = [1]
    for x in range(2, n + 1):
        multiply(x, res)
        
    res.reverse()
    return res

if __name__ == "__main__":
    assert "".join(map(str, factorial(5))) == "120"
    assert "".join(map(str, factorial(10))) == "3628800"
    print("Python: factorial passed tests!")
